### PR TITLE
CYOA - TweetMapper

### PR DESCRIPTION
**Pull request by dustinhansen**
For local deployment. A few extra features are recommended before wider deployment (see README).

### CODE REVIEW

File: arrowhead/arrowhead.rb [hunk 1]
```diff
@@ -15,16 +15,18 @@ class Arrowhead
     },
   }
 
-  # FIXME: I don't have time to deal with this.
+  def self.identify_shape(collection, shape)
```
Comment (djkotowski) [inline arrowhead/arrowhead.rb]: Extra file.

**Comment by djkotowski** [comment]
I am unable to do a complete review because of the number of files.  Use a CDN for external libraries (e.g. jQuery, Bootstrap, etc.) where possible, and ensure vendor CSS/JS that you must include is minified.

Please resubmit afterwards.

**Comment by djkotowski** [comment]
Delete unused directories.

File: cyoa/TweetMapper/app/models/tweet.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  validates :tweet_id_from_twitter, presence: true, numericality: true
+  validates :tweet_id_from_twitter, presence: true, uniqueness: true
+  validates_presence_of :author_screen_name
+  validates_presence_of :author_name
+  validates_presence_of :tweet_text
+  validates :latitude, presence: true, latitude: true
+  validates :longitude, presence: true, longitude: true
+  validates :tweet_created_at, presence: true, timeliness: true
+
+  class << self
```
Comment (djkotowski) [inline cyoa/TweetMapper/app/models/tweet.rb]: Prefer `def self.method` over `class << self` where possible.

File: cyoa/TweetMapper/app/models/tweet.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    def query_for_tweets(user_session)
+      @client.search('',
+        geocode: "#{user_session.latitude},#{user_session.longitude},#{user_session.miles_radius}mi",
+      ).take(Rails.configuration.tweets_to_take)
+    end
+
+    def upsert_tweet(tweet)
+      return unless tweet.geo?
+      Tweet.upsert({ tweet_id_from_twitter: tweet.id },
```
Comment (djkotowski) [inline cyoa/TweetMapper/app/models/tweet.rb]: Prefer `self` over class name in class methods.

Comment (djkotowski) [inline cyoa/TweetMapper/app/models/tweet.rb]: Also, wouldn't this make more sense as an instance method?

File: cyoa/TweetMapper/config/initializers/backtrace_silencers.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# Be sure to restart your server when you modify this file.
```
Comment (djkotowski) [inline cyoa/TweetMapper/config/initializers/backtrace_silencers.rb]: Delete unused files.

File: cyoa/TweetMapper/app/models/tweet.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
+require 'upsert/active_record_upsert'
```
Comment (djkotowski) [inline cyoa/TweetMapper/app/models/tweet.rb]: Typically, we use `find_or_create` over `upsert`.

File: cyoa/TweetMapper/app/models/tweet.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    end
+
+    private
+
+    def initialize_twitter_client
+      @client = Twitter::REST::Client.new do |config|
+        config.consumer_key = "BZAbS3emPSOQjXUAJM31q6nRk"
+        config.consumer_secret = "iMhoWfMBk5UEQnE0ESYVtpfk4drji5gXLbGydqzGuXVe3xjzHA"
+        config.access_token = "14140512-IbqG6CzfvVbRxvDp845B1wLOnJ4mIEBkNpXxpySx5"
+        config.access_token_secret = "Zq6uawBJ4aXImXzZTcWPdoAlUopTi9W6x108Xi7YCQ1i5"
```
Comment (djkotowski) [inline cyoa/TweetMapper/app/models/tweet.rb]: These keys should be in a secrets file, not in your public code.

File: cyoa/TweetMapper/app/assets/javascripts/application.js [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+//= require angular/angular
+//= require angular-google-maps/dist/angular-google-maps
+//= require lodash/lodash
+//= require jquery-timeago
+//= require twitter-text-js/js/twitter-text
+
+//= require_self
+//= require_tree .
+
+angular.module('TweetMapper', [
```
Comment (djkotowski) [inline cyoa/TweetMapper/app/assets/javascripts/application.js]: Avoid putting JS in the application.js file.  Store it in a separate JS file that you require here.

File: cyoa/TweetMapper/README.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+There's a few things we need to do before this can be deployed for mass usage:
+
+* Create additional Twitter application tokens and build out infrastructure for distributing requests among these various IDs (will allow us to squeeze more requests out of our tight rate limiting with Twitter).
+* Add intelligent throttling to the Twitter API calls to service all clients.
+
+Other nice-to-haves:
+
+* Add Twitter avatar images to infowindows.
+* Pull any linked images inline to have a preview showing in the infowindows.
```
Comment (djkotowski) [inline cyoa/TweetMapper/README.md]: Nice documentation.

File: cyoa/TweetMapper/vendor/assets/bower_components/bootstrap-sass/assets/stylesheets/bootstrap/mixins/_vendor-prefixes.scss [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+     -moz-hyphens: $mode;
+      -ms-hyphens: $mode; // IE10+
+       -o-hyphens: $mode;
+          hyphens: $mode;
+}
+
+// Placeholder text
+@mixin placeholder($color: $input-color-placeholder) {
+  // Firefox
+  &::-moz-placeholder {
```
Comment (djkotowski) [inline cyoa/TweetMapper/vendor/assets/bower_components/bootstrap-sass/assets/stylesheets/bootstrap/mixins/_vendor-prefixes.scss]: Except when creating a mixin to _avoid vendor prefixes_.

File: cyoa/TweetMapper/spec/models/user_session_spec.rb [hunk 10]
```diff
@@ -0,0 +1,31 @@
+require 'rails_helper'
+
+RSpec.describe UserSession, type: :model do
+  it 'has a valid factory' do
```
Comment (djkotowski) [inline cyoa/TweetMapper/spec/models/user_session_spec.rb]: [Linting](http://www.rubydoc.info/gems/factory_girl/file/GETTING_STARTED.md#Linting_Factories) provides this functionality.

**Comment by djkotowski** [comment]
A few comments throughout.  With the amount of reliance on JS and external APIs, there's not much you can test here.  Contract tests will help improve the reliability of your app.

File: cyoa/TweetMapper/app/assets/stylesheets/application.scss [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+.angular-google-map,
+.angular-google-map-container { height: 100%; }
+
+.infowindow {
+  .author_name {
+    font-size: 1.2em;
+    font-weight: bold;
+  }
+  .author_screen_name { }
```
Comment (paulh16) [inline cyoa/TweetMapper/app/assets/stylesheets/application.scss]: You can remove this and the next line as there are no styles.

File: cyoa/TweetMapper/app/controllers/tweets_controller.rb [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    json_response(session_errors, retrieve_latest_tweets)
+  rescue Twitter::Error::TooManyRequests
+    json_response(['Twitter rate limit exceeded. Please speak with a site administrator or wait a minute and try again.']);
+  end
+
+  private
+
+  def current_user_session
+    user_session = UserSession.find_by(session_key: session.id)
+    if user_session.nil?
```
Comment (paulh16) [inline cyoa/TweetMapper/app/controllers/tweets_controller.rb]: Consider extracting the logic in this if-else into a separate method. Is there a way to flip the logic of the if-else so you can use `if user_session`?

File: cyoa/TweetMapper/app/controllers/tweets_controller.rb [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+      user_session = UserSession.create(user_session_params)
+    else
+      user_session.update(user_session_params)
+    end
+    user_session
+  end
+
+  def retrieve_latest_tweets(limit = 100)
```
Comment (paulh16) [inline cyoa/TweetMapper/app/controllers/tweets_controller.rb]: Should this model logic belong in `Tweet.rb`?

File: cyoa/TweetMapper/app/controllers/tweets_controller.rb [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+  def latest
+    session_errors = parse_session_errors(current_user_session)
+    return json_response(session_errors) if session_errors.length > 0
+    Tweet.harvest_from_twitter
+    json_response(session_errors, retrieve_latest_tweets)
+  rescue Twitter::Error::TooManyRequests
+    json_response(['Twitter rate limit exceeded. Please speak with a site administrator or wait a minute and try again.']);
+  end
+
+  private
```
Comment (paulh16) [inline cyoa/TweetMapper/app/controllers/tweets_controller.rb]: Nice job making the methods below private.

File: cyoa/TweetMapper/app/models/user_session.rb [hunk 15]
```diff
@@ -0,0 +1,11 @@
+require 'upsert/active_record_upsert'
+
+class UserSession < ActiveRecord::Base
+  validates :session_key, presence: true, uniqueness: true
+  validates :last_activity, presence: true, timeliness: true
+  validates :latitude, presence: true, numericality: true
+  validates :longitude, presence: true, numericality: true
+  validates :miles_radius, presence: true, numericality: true
+
+  before_validation { self.last_activity = Time.now }
```
Comment (paulh16) [inline cyoa/TweetMapper/app/models/user_session.rb]: Excellent callback.

File: cyoa/TweetMapper/db/migrate/20150623202819_create_tweets.rb [hunk 16]
```diff
@@ -0,0 +1,14 @@
+class CreateTweets < ActiveRecord::Migration
+  def change
+    create_table :tweets do |t|
+      t.decimal :tweet_id
```
Comment (paulh16) [inline cyoa/TweetMapper/db/migrate/20150623202819_create_tweets.rb]: Is there a reason to manually add the id to this model? Rails usually does this automatically?

File: cyoa/TweetMapper/db/migrate/20150623202819_create_tweets.rb [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+      t.string :author
+      t.string :tweet, limit: 140
+      t.integer :retweet_count
+      t.float :latitude, precision: 10, scale: 6
+      t.float :longitude, precision: 10, scale: 6
+      t.datetime :created_at
+      t.timestamps null: false
```
Comment (paulh16) [inline cyoa/TweetMapper/db/migrate/20150623202819_create_tweets.rb]: Is there a reason to have a column named `datetime`? I believe this is created with the `timestamps` method.

File: cyoa/TweetMapper/db/migrate/20150623202819_create_tweets.rb [hunk 18]
```diff
@@ -0,0 +1,14 @@
+class CreateTweets < ActiveRecord::Migration
```
Comment (paulh16) [inline cyoa/TweetMapper/db/migrate/20150623202819_create_tweets.rb]: Nice job adding constraints to the migration.

File: cyoa/TweetMapper/spec/factories/tweets.rb [hunk 19]
```diff
@@ -0,0 +1,15 @@
+require 'faker'
```
Comment (paulh16) [inline cyoa/TweetMapper/spec/factories/tweets.rb]: Excellent use of Faker!

File: cyoa/TweetMapper/spec/models/tweet_spec.rb [hunk 20]
```diff
@@ -0,0 +1,39 @@
+require 'rails_helper'
+
+RSpec.describe Tweet, type: :model do
+  it 'has a valid factory' do
+    tweet = create(:tweet)
+    expect(tweet).to be_valid
+  end
```
Comment (paulh16) [inline cyoa/TweetMapper/spec/models/tweet_spec.rb]: I'd add line breaks between these specs.

**Comment by paulh16** [comment]
Overall, this is impressive! I just want to pose the question about whether more controller specs would be effective. You have more complexity in your controllers than the typical Rails app, so I do feel that testing the various pathways in your controllers would be good.

Nice work!

**Pull request closed without merging**